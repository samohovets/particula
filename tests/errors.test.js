/* eslint-env jest */
const path = require('path');
const express = require('express');
const request = require('supertest');
const {setupErrorHandling} = require('../src/express');

test('Should handle 404', async done => {
  const app = express();
  await setupErrorHandling(app);

  await request(app)
    .get('/404')
    .expect(404, 'Route /404 Not found.');

  done();
});

test('Should handle 500 errors in development', async done => {
  const app = express();
  app.get('/', () => {
    throw new Error('test error');
  });
  await setupErrorHandling(app);

  const {text} = await request(app)
    .get('/')
    .expect(500);

  // replace local paths
  const currentPath = path.resolve(__dirname).replace('/tests', '');
  const pathRegex = new RegExp(currentPath, 'g');

  // replace possible mismatching stack traces
  const stack1 = /at Server.emit \(events.js:.+?\)/g;
  const stack2 = /at parserOnIncoming \(_http_server.js:.+?\)/g;
  const stack3 = /at HTTPParser.parserOnHeadersComplete \(_http_common.js:.+?\)/g;

  // replace all things
  const cleanText = text
    .replace(pathRegex, '')
    .replace(stack1, '')
    .replace(stack2, '')
    .replace(stack3, '');

  expect(cleanText).toMatchSnapshot();

  done();
});

test('Should handle 500 errors in production', async done => {
  process.env.NODE_ENV = 'production';
  const app = express();
  app.get('/', () => {
    throw new Error('test error');
  });
  await setupErrorHandling(app);

  await request(app)
    .get('/')
    .expect(500, '500: Internal Server Error');

  done();
});
