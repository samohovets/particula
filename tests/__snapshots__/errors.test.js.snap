// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should handle 500 errors in development 1`] = `
"<html>
  <head>
    <meta charset='utf-8'> 
    <title>Error: test error</title>
    <style>* {
  margin: 0;
  padding: 0;
  outline: 0;
}

body {
  padding: 80px 100px;
  font: 13px \\"Helvetica Neue\\", \\"Lucida Grande\\", \\"Arial\\";
  background: #ECE9E9 -webkit-gradient(linear, 0% 0%, 0% 100%, from(#fff), to(#ECE9E9));
  background: #ECE9E9 -moz-linear-gradient(top, #fff, #ECE9E9);
  background-repeat: no-repeat;
  color: #555;
  -webkit-font-smoothing: antialiased;
}
h1, h2 {
  font-size: 22px;
  color: #343434;
}
h1 em, h2 em {
  padding: 0 5px;
  font-weight: normal;
}
h1 {
  font-size: 60px;
}
h2 {
  margin-top: 10px;
}
ul li {
  list-style: none;
}
#stacktrace {
  margin-left: 60px;
}
</style>
  </head>
  <body>
    <div id=\\"wrapper\\">
      <h1>Connect</h1>
      <h2><em>500</em> Error: test error</h2>
      <ul id=\\"stacktrace\\"><li> &nbsp; &nbsp;at app.get (/tests/errors.test.js:21:11)</li><li> &nbsp; &nbsp;at Layer.handle [as handle_request] (/node_modules/express/lib/router/layer.js:95:5)</li><li> &nbsp; &nbsp;at next (/node_modules/express/lib/router/route.js:137:13)</li><li> &nbsp; &nbsp;at Route.dispatch (/node_modules/express/lib/router/route.js:112:3)</li><li> &nbsp; &nbsp;at Layer.handle [as handle_request] (/node_modules/express/lib/router/layer.js:95:5)</li><li> &nbsp; &nbsp;at /node_modules/express/lib/router/index.js:281:22</li><li> &nbsp; &nbsp;at Function.process_params (/node_modules/express/lib/router/index.js:335:12)</li><li> &nbsp; &nbsp;at next (/node_modules/express/lib/router/index.js:275:10)</li><li> &nbsp; &nbsp;at expressInit (/node_modules/express/lib/middleware/init.js:40:5)</li><li> &nbsp; &nbsp;at Layer.handle [as handle_request] (/node_modules/express/lib/router/layer.js:95:5)</li><li> &nbsp; &nbsp;at trim_prefix (/node_modules/express/lib/router/index.js:317:13)</li><li> &nbsp; &nbsp;at /node_modules/express/lib/router/index.js:284:7</li><li> &nbsp; &nbsp;at Function.process_params (/node_modules/express/lib/router/index.js:335:12)</li><li> &nbsp; &nbsp;at next (/node_modules/express/lib/router/index.js:275:10)</li><li> &nbsp; &nbsp;at query (/node_modules/express/lib/middleware/query.js:45:5)</li><li> &nbsp; &nbsp;at Layer.handle [as handle_request] (/node_modules/express/lib/router/layer.js:95:5)</li><li> &nbsp; &nbsp;at trim_prefix (/node_modules/express/lib/router/index.js:317:13)</li><li> &nbsp; &nbsp;at /node_modules/express/lib/router/index.js:284:7</li><li> &nbsp; &nbsp;at Function.process_params (/node_modules/express/lib/router/index.js:335:12)</li><li> &nbsp; &nbsp;at next (/node_modules/express/lib/router/index.js:275:10)</li><li> &nbsp; &nbsp;at Function.handle (/node_modules/express/lib/router/index.js:174:3)</li><li> &nbsp; &nbsp;at Function.handle (/node_modules/express/lib/application.js:174:10)</li><li> &nbsp; &nbsp;at Server.app (/node_modules/express/lib/express.js:39:9)</li><li> &nbsp; &nbsp;</li><li> &nbsp; &nbsp;</li><li> &nbsp; &nbsp;</li></ul>
    </div>
  </body>
</html>
"
`;
